### ISSUE TITLE
3.0.0 release

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#902]
Can be labelled **ready** when all other issues on this milestone are closed.

- [x] check integration package compatibility
  - [x] [AutoFixture.AutoFakeItEasy](https://github.com/AutoFixture/AutoFixture) - @blairconrad 
    - fails - can't find `IFakeOptionsBuilder<T>` in the new FakeItEasy. This is no different than the case when using AutoFixture.AutoFakeItEasy against FakeItEasy 2.x. Fixing AutoFixture/AutoFixture#693 would resolve this, but AutoFixture currently [suffers from interregnum](https://github.com/AutoFixture/AutoFixture/issues/703), so there's no reason to expect any change soon.
    - Related issue: https://github.com/AutoFixture/AutoFixture/issues/693
  - [x] [AutoFixture.AutoFakeItEasy2](https://github.com/AutoFixture/AutoFixture)
    - no issues. :shipit: 
  - [x] [Machine.Fakes.FakeItEasy](https://github.com/machine/machine.fakes) - @adamralph
    - Raised https://github.com/machine/machine.fakes/issues/37
  - [x] [Autofac.Extras.FakeItEasy](https://github.com/autofac/Autofac.Extras.FakeItEasy) - @thomaslevesque 
    - Currently restricted to FIE `[2.0.0,3.0.0)`, so releasing 3.0.0 won't break it.
    - One test fails when upgrading to FIE 3.0.0. It relies on the old behavior of `DoesNothing()` that caused the method to return null, but now it returns a dummy instead. `AutoFake` has an optional `callsDoNothing` parameter which causes all methods to be configured with `DoesNothing()` when the fake is created, but doing this on a new fake has no effect for most methods (exceptions: ToString, GetHashCode, Equals, properties), it just resets them to the default behavior. It would probably make sense to remove the `callsDoNothing` parameter.
    - Related issue: https://github.com/autofac/Autofac.Extras.FakeItEasy/issues/10
  - [x] [EntityFrameworkTesting.FakeItEasy](https://github.com/scott-xu/EntityFramework.Testing) - @blairconrad 
    - Fails. `IReturnValueConfiguration.ReturnsLazily` signature changed. On the other hand, the package is restricted to >= 1.25.2 && < 2.0.0, so I don't think we need to worry about it. 
    - Related issue: https://github.com/scott-xu/EntityFramework.Testing/issues/17
- [x] run code analysis in VS in *Release* mode and address violations (send a regular PR which must be merged before continuing)
- [x] if necessary, change `VERSION_SUFFIX` in [AppVeyor](https://ci.appveyor.com/project/FakeItEasy/fakeiteasy/settings) to appropriate "-beta123" or "" (for non-betas) value and initiate a build
- [x] check build
- edit draft release in [GitHub UI](https://github.com/FakeItEasy/FakeItEasy/releases):
    - [x] ensure completeness of release notes, including non-owner contributors, if any (move release notes forward from any pre-releases to the current release)
    - [x] attach main nupkg and/or analyzer nupkg, whichever have content to release
    - [x] publish the release
- [x] push nupkgs to NuGet using the Deploy option in AppVeyor
- [x] de-list pre-release or superseded buggy NuGet packages if present
- [x] tweet, mentioning contributors and post link as comment here for easy retweeting ;-)
- [x] post tweet in [Gitter](https://gitter.im/FakeItEasy/FakeItEasy)
- [x] post a link to the GitHub Release in each issue in this milestone, with thanks to contributors
- [x] run `rake next_version[new_version]` to
    - create a pull request that changes the version in CommonAssemblyInfo.cs to the expected version (of form _xx.yy.zz_)
    - create a new draft GitHub Release
    - create a new milestone for the next release
    - create a new issue (like this one) for the next release, adding it to the new milestone
- [x] close this milestone

**Comment by blairconrad** [comment]
1. How long do we want to keep an RC available before we release suffixless? A week? :grinning: 
   We could aim for the 22nd.
2. Do we want to bother testing the 3 frameworks that we did when we released 2.0? AutoFixture, AutoFac, and Machine.Fakes? I don't mind. I can take AutoFixture again, but may not get it tested 'til the weekend.

**Comment by thomaslevesque** [comment]
> How long do we want to keep an RC available before we release suffixless? A week? 😀
> We could aim for the 22nd.

👍 

> Do we want to bother testing the 3 frameworks that we did when we released 2.0? AutoFixture, AutoFac, and Machine.Fakes? I don't mind. I can take AutoFixture again, but may not get it tested 'til the weekend.

We probably should. The changes in the surface API probably won't affect end users, but third party frameworks are more likely to depend on the exact configuration interfaces.

**Comment by adamralph** [comment]
👍 to both points.

**Comment by blairconrad** [comment]
Ah. Someone updated the checklist. Thanks, someone!

**Comment by adamralph** [comment]
That was me, with placeholders to fill in, similar to how we did it in #336. 😃 

I don't suppose any other significant dependent packages have appeared in the meantime?

**Comment by thomaslevesque** [comment]
I just checked `Autofac.Extras.FakeItEasy`. No build error, but one test fails due to #830 

https://github.com/autofac/Autofac.Extras.FakeItEasy/blob/develop/test/Autofac.Extras.FakeItEasy.Test/AutoFakeFixture.cs#L92-L101

`callsDoNothing: true` causes all calls to be configured with `DoesNothing()`. If they want to keep the same behavior as before, they'll have to configure the calls to return the default value instead. I'll create an issue on their repo.

**Comment by thomaslevesque** [comment]
> I'll create an issue on their repo.

In fact, there's already an issue for upgrading to FIE 3.0.0: https://github.com/autofac/Autofac.Extras.FakeItEasy/issues/10

**Comment by adamralph** [comment]
The Autofac.Extras.FakeItEasy dependency range is set to FakeItEasy (>= 2.0.0 && < 3.0.0) so there's nothing to do immediately, other than perhaps warn them that they'll have to address https://github.com/autofac/Autofac.Extras.FakeItEasy/issues/10 soon.

**Comment by adamralph** [comment]
I trimmed my comment down.

**Comment by thomaslevesque** [comment]
> Sorry @thomaslevesque, I raced you to comment on autofac/Autofac.Extras.FakeItEasy#10 - I should have left it to you.

No problem :wink:

**Comment by godrose** [comment]
Does this release support the latest Castle.Core version? If it has a
dependency on it, of course...

**Comment by blairconrad** [comment]
@godrose, yes the 3.0.0 release will use Castle.Core 4.0.0.

**Comment by blairconrad** [comment]
> That was me, with placeholders to fill in, similar to how we did it in #336. 😃

Yeah, I was heading here from there with my own copy on the clipboard!

> I don't suppose any other significant dependent packages have appeared in the meantime?

Um, [fakexrmeasy](http://stackoverflow.com/questions/42255941/creating-linked-data-with-fakexrmeasy)?
No, but seriously, [EntityFrameworkTesting.FakeItEasy](https://www.nuget.org/packages/EntityFrameworkTesting.FakeItEasy) has about as many downloads per day as the others (perhaps more than Machine.Fakes.FakeItEasy).

**Comment by adamralph** [comment]
It sounds like EntityFrameworkTesting.FakeItEasy is worth adding to the list.

**Comment by blairconrad** [comment]
Done!

**Comment by thomaslevesque** [comment]
> Um, fakexrmeasy?

Ah, I guess we saw the same Stackoverflow question... This framework is using FIE 1.25.3, so I guess they're not interested in upgrading.

> No, but seriously, EntityFrameworkTesting.FakeItEasy has about as many downloads per day as the others (perhaps more than Machine.Fakes.FakeItEasy).

I don't expect any issue with it. It just contains one extension method, which seems pretty straightforward: https://github.com/scott-xu/EntityFramework.Testing/blob/master/src/EntityFramework.Testing.FakeItEasy/FakeItEasyDbSetExtensions.cs (plus a small bit of code for EntityFrameworkTesting.FakeItEasy.NInject)

**Comment by blairconrad** [comment]
> This framework is using FIE 1.25.3, so I guess they're not interested in upgrading.

AutoFixture is using 1.7.4257.42, so I can't blame EntityFrameworkTesting.FakeItEasy!

**Comment by adamralph** [comment]
> This framework is using FIE 1.25.3, so I guess they're not interested in upgrading.

Depending on a old version shouldn't be regarded negatively. The older the minimum dependency, the larger the potential user base for the library, so making the dependency as old as possible is the right thing to do. That said, the library should be tested with newer versions. At the minimum it should be tested with each subsequent major version.

**Comment by blairconrad** [comment]
>  the library should be tested with newer versions. At the minimum it should be tested with each subsequent major version.

Agreed. Just not necessarily by us. I think the 4 we've committed to (5 if we count AutoFixture.AutoFakeItEasy and AutoFixture.AutoFakeItEasy2) is enough.

**Comment by adamralph** [comment]
I agree. We have the main ones covered.

**Comment by thomaslevesque** [comment]
> AutoFixture is using 1.7.4257.42, so I can't blame EntityFrameworkTesting.FakeItEasy!

Actually I was talking about fakexrmeasy... I haven't check which version EntityFrameworkTesting.FakeItEasy is using

**Comment by blairconrad** [comment]
Oh, for the love of Pete! I was trying to talk about fakexrmeasy as well. Sigh.

**Comment by blairconrad** [comment]
I gotta start reading what I paste.

**Comment by blairconrad** [comment]
EntityFrameworkTesting.FakeItEasy fails. `IReturnValueConfiguration.ReturnsLazily` signature changed. On the other hand, the package is restricted to >= 1.25.2 && < 2.0.0, so I don't think we need to worry about it. 
I've checked the "done" box way upstairs.

**Comment by blairconrad** [comment]
@thomaslevesque said

> I just checked `Autofac.Extras.FakeItEasy`. No build error, but one test fails due to #830 

The fact that you're recompiling worries me. That will test for one kind of problem, but not all. Maybe I misunderstand what you're doing, but consider https://github.com/AutoFixture/AutoFixture/issues/693#issuecomment-244396725. Sometimes a recompile will work, but the underlying types have changed in a way that makes previously-compiled assemblies incompatible with the new ones. I think updating FakeItEasy and adding redirects is the better way to check.

**Comment by adamralph** [comment]
@blairconrad should we raise an issue for EFT.FIE to upgrade to 3.0.0?

**Comment by adamralph** [comment]
@thomaslevesque would you mind writing up your findings against Autofac.Extras.FakeItEasy in the checklist?

Who checked  AutoFixture.AutoFakeItEasy? Is it worth raising an issue there to upgrade to FIE 3.0.0?

**Comment by blairconrad** [comment]
@adamralph, I checked the AutoFixtures.
I don't think it's worth raising an issue. As far as I can tell, if they just merged AutoFixture/AutoFixture#702 to fix AutoFixture/AutoFixture#693, everything would work. And they seem to be happy with my PR. It's just that without a governance model, nothing's being done over there. I think a new issue would just add to the pile.

> should we raise an issue for EFT.FIE to upgrade to 3.0.0?

I don't think so. They seem to have been happy with < 2 up 'til now. When they _do_ decide to open the floodgates, they'll (I assume) recompile anyhow. When I recompile with 3.0.0-rc001, their EntityFramework.Testing.FakeItEasy.Tests all pass.

Also they have another project, EntityFramework.Testing.FakeItEasy.Ninject, that depends on Ninject.MockingKernel.FakeItEasy, which has the same FakeItEasy version range restriction as EntityFramework.Testing.FakeItEasy. I don't want to start a whole big thing.

**Comment by thomaslevesque** [comment]
> The fact that you're recompiling worries me. That will test for one kind of problem, but not all. Maybe I misunderstand what you're doing, but consider AutoFixture/AutoFixture#693 (comment). Sometimes a recompile will work, but the underlying types have changed in a way that makes previously-compiled assemblies incompatible with the new ones. I think updating FakeItEasy and adding redirects is the better way to check.

Ah, yes, I was checking source compatibility, not binary compatibility.

AutoFac.Extras.FakeItEasy is restricted to FIE `[2.0.0,3.0.0)`, so it won't accidentally update to 3.0.0+. And anyway they intend to upgrade to FIE 3.0.0, so it will be recompiled.

**Comment by thomaslevesque** [comment]
> @thomaslevesque would you mind writing up your findings against Autofac.Extras.FakeItEasy in the checklist?

Done

**Comment by adamralph** [comment]
I added links, names and issues to all the integration packages. I think we have them covered now and we can go ahead with the rest of the release when we're ready.

**Comment by blairconrad** [comment]
Nice work, @adamralph!

**Comment by blairconrad** [comment]
https://twitter.com/fakeiteasyfx/status/834359200826392577

https://twitter.com/fakeiteasyfx/status/834359146577293313

**Comment by blairconrad** [comment]
I'm closing [3.0.0-pre](https://github.com/FakeItEasy/FakeItEasy/milestone/30) too.
